android-sdk/README.md

```diff
@@ -0,0 +1,224 @@
+
+### Implementation notes
+
+Public APIs should always expose native Java Objects and not stringified JSON or JSON Objects.
+
+### Use cases and psuedo code examples
+
+#### Starting up Mobile Core
+MobileCore is the "main" class for the SDK of the Core SDK.  Users instanciate it by placing a client configuration file in their assets directory, configuring MobileCore.Builder, and then calling its build method.
+
+```
+MobileCore.Builder builder = new MobileCore.Builder(application);
+MobileCore core = builder.build();
+```
+
+During the build method, MobileCore will scan a registry of services and configure them by calling the method `bootstrap` defined in the ServiceModule interface.  MobileCore will pass to bootstrap the core instance and the section of mobile-core.json that corresponds to that service.
+
+#### Writing a Service Module
+A service interface will be responsible for building requests, parsing responses, handling errors.  Services may depend on each other and the method of service composition is currently under evaluation. However, services will have a resolution order where a request is passed along a chain until it is ready to be executed.  Mobile Core will provide the services `threading`, `logging`, and `networking`.  ServiceModules will implement the service module class and provide a default no argument constructor.  
+
+Example KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+   /* snip */
+   public KeyCloakService() {}
+
+   @Override
+    public void bootstrap(MobileCore core, ServiceConfiguration config) {
+        this.serverUrl = config.getProperty("auth-server-url");
+        this.clientId = config.getProperty("client_id");
+        this.audience = config.getProperty("audience");
+        this.grantType = config.getProperty("grant_type");
+        this.subjectTokenType = config.getProperty("subject_token_type");
+        this.requestedTokenType = config.getProperty("requested_token_type");
+        this.realm = config.getProperty("realm");
+        this.core = core;
+    }
+}
+```
+
+#### Registering a Service Module and the ServiceModuleRegistry
+The Core SDK will provide a service registry that modules can register themselves with as well as register their dependencies.  During build MobileCore will call bootstrap on services in the order of those with the least dependencies to those with the most dependencies.
+
+Example registration of KeyCloakService
+```
+public class KeyCloakService implements ServiceModule {
+  static {
+    ServiceModuleRegistry.registerServiceModule("keycloak", KeyCloakService.class, "http");
```

Comment:
ServiceModuleRegistry has static method (as mentioned above we should avoid that) so we can make it singleton